init: handle empty settings file gracefully

furszy opened the pull request:
Small and simple issue reported [here](https://community.umbrel.com/t/bitcoin-docker-container-keeps-restarting/2144).

Improving a confusing situation reported by users who did not understand why a 
settings parsing error occurred when the file was empty and did not know how to solve it.

Empty setting file could be due (1) corruption or (2) an user manually cleaning up the file content.
In both scenarios, the 'Unable to parse settings file' error does not help the user move forward.

0xB10C commented:
see previous discussion in #23096

furszy reviewed (commented):
> see previous discussion in #23096, #22591, and [#21340 (comment)](https://github.com/bitcoin/bitcoin/issues/21340#issuecomment-880147010)

Hmm, okay, agree. Thanks.
It would probably be useful to introduce support for comments. This way, we can write something at the beginning of the file, ensuring that users and other software developers don't clean it up manually, thinking that it will be regenerated automatically. Do you know if something like this has been proposed before? @0xB10C.

shaavan reviewed (commented):
I agree with the reasoning ryanofsky has given [here](https://github.com/bitcoin/bitcoin/pull/23096#issuecomment-927785519).

> A zero-size settings file is a corrupt settings file.

But going through the [issue](https://community.umbrel.com/t/bitcoin-docker-container-keeps-restarting/2144) mentioned in the description. I don't think `Unable to parse settings file /data/.bitcoin/settings.json` is a very helpful message when it comes to bitcoind detecting an empty settings.json file.

I think we should still handle the case of empty file separately. But instead of passing the file as valid, we should rather print a very clear error message, pointing the user to the reason for failure and possible remedy. This will definitely better the User experience for the noderunners.

furszy reviewed (commented):
> I think we should still handle the case of empty file separately. But instead of passing the file as valid, we should rather print a very clear error message, pointing the user to the reason for failure and possible remedy. This will definitely better the User experience for the noderunners.

For GUI interactive users, this isn't an issue. We show a dialog that allows continuing when a parsing error is found (still, we might want to be clearer about the fact that they could be disabling Tor and/or other important configurations by allowing this).

For non-interactive ones, people or other softwares, I'm inclined to solve this situation by adding a comment at the top of the file. Something like "This is an auto-generated JSON-formatted file. Please refrain from modifying it unless you are certain about the potential consequences."

anibilthare commented on src/common/settings.cpp:
```diff
@@ -80,7 +80,9 @@ bool ReadSettings(const fs::path& path, std::map<std::string, SettingsValue>& va
+    std::string content{std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>()};
+    if (content.empty()) return true; // Ok for empty file
+    if (!in.read(content)) {
         errors.emplace_back(strprintf("Unable to parse settings file %s", fs::PathToString(path)));
```
I agree with @shaavan  [here](https://github.com/bitcoin/bitcoin/pull/29144#pullrequestreview-1798049445) . What if we simply modify the error message here and keep rest of the code as is to be more user friendly. Referring to https://github.com/bitcoin/bitcoin/pull/23096, https://github.com/bitcoin/bitcoin/pull/22591, and https://github.com/bitcoin/bitcoin/issues/21340#issuecomment-880147010 as mentioned earlier, following change can be added to the error message
```suggestion
        errors.emplace_back(strprintf("Unable to parse settings file %s, the file may be corrupt or empty(Consider deleting the file if it is empty). If the settings file is getting truncated automatically, check if there is suffiecient disk space available.", fs::PathToString(path)));
```

shaavan replied:
I see your point here. But wording like this provides a very unclear reason for failure.
Don't you think we would be throwing users in too many directions here?

anibilthare replied:
Can you please point out exactly what part of these statements might confuse the user?
Although we'll be asking the user to check for 2 things in the comment but I believe they are pretty straight forward, right?

1. Is there any space left on the disk.
2. Is settings.json empty or not?

If none of the above conditions hold true and user is still seeing the issue then we are indicating the possibility of corruption of the settings file itself.

1440000bytes replied:
NACK for this PR

Core does not works for umbrel

This is external and need to maintain outside of core: https://community.umbrel.com/t/bitcoin-docker-container-keeps-restarting/2144

furszy replied:
> Core does not works for umbrel

@1440000bytes, I did not know about umbrel existence until you pointed it out.

The PR tries to improve a confusing situation reported by users who did not understand why a settings parsing error occurred when the file was empty and did not know how to solve it.

shaavan reviewed (commented):
I get your point, @furszy, and I can see your approach helping with smoother handling of the "file empty" situation.

However, pointing back again to ryanofsky comment

> A zero-size settings file is a corrupt settings file.

The interpretation I draw from this is if a file is empty (and hence corrupted once) there is no telling what could be the reason for that happening or whether such corruption could happen again.
So, I believe being very critical and clear with the handling will be better for the following reasons:

1. For crucial backend software, such as bitcoind, it seems negligent to not properly inform the user that their settings file has been lost to corruption.
2. It seems to me that a regular user won't tinker with their settings file regularly, so any kind of auto-generated message in the settings.json file loses its relevance there.
3. Finally, say a user sets some custom settings once and later loses the settings to corruption (=> file now empty); it seems an improper approach to me for bitcoind to return to default settings without any kind of proper log or error to the user.

Also, a gentle ping to @ryanofsky to confirm if my interpretation of the situation is correct, and also to get his take on this PR.

kristapsk commented:
> It would probably be useful to introduce support for comments. This way, we can write something at the beginning of the file, ensuring that users and other software developers don't clean it up manually, thinking that it will be regenerated automatically. 

This will not help if cause was full disk. https://github.com/bitcoin/bitcoin/issues/21340#issuecomment-876779561

shaavan reviewed (commented):
> Thus I proposed a different solution for non-interactive users who are manually modifying the file. See https://github.com/bitcoin/bitcoin/pull/29144#pullrequestreview-1798109497.
Stating that the file is JSON-formatted and auto-generated gives bitcoind users another hint over what they can and can't do.

I see your point, @furszy 

But I think my points still apply to creating an "auto-generated" file.

1. A regular user might not tinker with their settings file regularly and hence the "this file is auto-generated" might go unnoticed by the user essentially making the message irrelevant.

2. If a user had a _custom_ set of settings and later loses them to corruption, they might not know that this has happened for a while if they are not given an "on-the-face" alert. And for critical software such as bitcoind, this seems to me a non-optimal behavior.

I think it's a great idea to help the user with an auto-generated file in case of setting loss. But in the absence of an "on-the-face" alert, that change loses much of its significance.

ryanofsky commented:
re: 725a1fc7a7dbb1153bec188eda2f17a217560352 I think it is important to treat an empty settings file as an error, because the most likely cause of an empty file is some kind of crash or corruption or disk full state, and it is safer to alert the user about the situation so they can fix it, than proceed as if the error didn't happen and lose important settings that could affect security or stability.

I think the error message "Unable to parse settings file %s" could definitely be improved though. Would suggest maybe "Settings file %s does not contain valid JSON. This is probably caused by disk corruption or a crash, and can be fixed by just removing the file, which will reset settings to default values." This would probably work OK for both command line and GUI (since the GUI already offers the option to reset the settings see https://github.com/bitcoin-core/gui/pull/379).

As far as adding a comments to the settings file, I'm not sure it would necessarily help in the reported case from https://community.umbrel.com/t/bitcoin-docker-container-keeps-restarting/2144, but it probably depends on whatever caused the settings file to be truncated or created empty in that case, and it could be a good idea to add comments to the file regardless. JSON doesn't have a syntax for comments, it but would be possible to add an extra field like:

```json
{
  "_comment_": "This file is automatically generated and updated by Bitcoin Core. Please do not edit this file while the node is running, as any changes might be ignored or overwritten."
}
```

furszy committed:
init: improve corrupted/empty settings file error msg

The preceding "Unable to parse settings file" message lacked
the necessary detail and guidance for users on what steps to
take next in order to resolve the startup error.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

furszy committed:
settings: add auto-generated warning msg for editing the file manually

Hopefully, refraining users from modifying the file unless they are
certain about the potential consequences.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

furszy commented:
> I think the error message "Unable to parse settings file %s" could definitely be improved though. Would suggest maybe "Settings file %s does not contain valid JSON. This is probably caused by disk corruption or a crash, and can be fixed by removing the file, which will reset settings to default values." This would probably work OK for both command line and GUI (since the GUI already offers the option to reset the settings see [bitcoin-core/gui#379](https://github.com/bitcoin-core/gui/pull/379)).
> 
> As far as adding a comments to the settings file, I'm not sure it would necessarily help in the reported case from https://community.umbrel.com/t/bitcoin-docker-container-keeps-restarting/2144, but it probably depends on whatever caused the settings file to be truncated or created empty in that case, and it could be a good idea to add comments to the file regardless. JSON doesn't have a syntax for comments, it but would be possible to add an extra field like:
> 
> ```json
> {
>   "_comment_": "This file is automatically generated and updated by Bitcoin Core. Please do not edit this file while the node is running, as any changes might be ignored or overwritten."
> }
> ```

Sounds great. Thanks @ryanofsky!
Updated per feedback.

furszy reviewed (commented):
Cool notification. The CI should be happy now.

ryanofsky reviewed (approved):
Code review ACK 6f3183e0f0dec375241db96db417aa26e7c3e987. This seems like a helpful change.

But It does seem unfortunate that both of the commits here make tests more verbose and repetitive. And the tests even before this PR were probably unnecessarily fragile. But I'm not sure I see a better way to write the tests, or a way to avoid the repetition of warning and error messages. If anybody has any ideas on how to improve the tests here, I'd be interested to know.

furszy commented on test/functional/feature_settings.py:
```diff
@@ -53,7 +53,7 @@ def run_test(self):
         with settings.open("w") as fp:
             fp.write("invalid json")
-        node.assert_start_raises_init_error(expected_msg='Unable to parse settings file', match=ErrorMatch.PARTIAL_REGEX)
+        node.assert_start_raises_init_error(expected_msg=f'does not contain valid JSON. This is probably caused by disk corruption or a crash, and can be fixed by removing the file, which will reset settings to default values.', match=ErrorMatch.PARTIAL_REGEX)
```
yeah. Leftover from a previous implementation. I was checking the complete error message, including the settings file path (which was making the Windows CI job angry).

shaavan reviewed (commented):
Thanks for the update, @furszy
The new error message and the settings.json warnings are very precise in their details and I believe will be better in guiding the user in case of error, and refraining them from modifying settings without being cautious.

I just have one suggestion I want to add.

ryanofsky commented on src/qt/test/optiontests.cpp:
```diff
@@ -61,7 +61,10 @@ void OptionTests::migrateSettings()
     QVERIFY(!settings.contains("addrSeparateProxyTor"));
 
     std::ifstream file(gArgs.GetDataDirNet() / "settings.json");
+    std::string default_warning = strprintf("This file is automatically generated and updated by %s. Please do not edit this file while the node "
+                                            "is running, as any changes might be ignored or overwritten.", PACKAGE_NAME);
```
I think in general it is safe to edit the file, and main thing worth pointing out is when changes might be ignored or overwritten. It's possible adding more information could be useful, but I don't think there's anything clearly missing. The commit message also seems ok and wouldn't really be an appropriate place to caution users since they are unlikely to see it (in the file itself or in documentation would be better).

furszy replied:
Since the confusion over the empty settings file error is resolved by the first commit. I agree with ryanofsky; while extending the file warning message sounds plausible, I'm also ok with the current warning.

furszy reviewed (commented):
> But It does seem unfortunate that both of the commits here make tests more verbose and repetitive. And the tests even before this PR were probably unnecessarily fragile. But I'm not sure I see a better way to write the tests, or a way to avoid the repetition of warning and error messages. If anybody has any ideas on how to improve the tests here, I'd be interested to know.

This is not a small change, but.. we could work on a `Resources` class that is auto-generated during the compile phase. It would store all static error and warning message strings in conjunction with an auto-generated ID for each of them. Allowing us to access messages via their IDs instead of using the strings directly. The ID names could follow a certain layering pattern; e.g. accessing a init error could be: `R.getStr(R.errors.init.settings_invalid_json)`. (similar to how Android organizes app resources).

Then, to prevent code and string repetition inside functional and unit tests, the Resources class generation script could output C++ and Python compatible code.

@ryanofsky, just an idea. Not sure if it worth the efforts.

furszy reviewed (commented):
Rebased to cleanup spurious CI failure.

hebasto reviewed (approved):
ACK d664eab66ecfa1d318b11de5deafb04b1e3c4c06, tested on Ubuntu 23.10.

nano-nit: There are some complaints from [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy).

furszy commented:
> nano-nit: There are some complaints from [clang-format-diff.py](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy).

Done.

hebasto reviewed (approved):
re-ACK e9014042a6bed8c16cc9a31fc35cb709d4b3c766.

ryanofsky reviewed (approved):
Code review ACK e9014042a6bed8c16cc9a31fc35cb709d4b3c766. Just whitespace formatting changes and shortening a test string literal since last review

shaavan reviewed (commented):
Code review ACK e9014042a6bed8c16cc9a31fc35cb709d4b3c766

After our chat in [this](https://github.com/bitcoin/bitcoin/pull/29144#discussion_r1450404094) thread, I believe this PR is good to go!

There are just a few whitespace tweaks since my last comment, and they look great too.

luke-jr commented:
Should we perhaps add (and recommend, for non-GUI users), a `-resetguisettings` for non-GUI?

achow101 commented:
ACK e9014042a6bed8c16cc9a31fc35cb709d4b3c766

furszy reviewed (commented):
> Should we perhaps add (and recommend, for non-GUI users), a `-resetguisettings` for non-GUI?

Sure. We could introduce a `-resetsettings` in a follow-up.

fanquake commented:
So now we will (always?) log this on startup:
```bash
2024-01-23T22:48:12.205684Z [init] Setting file arg: _warning_ = "This file is automatically generated and updated by Bitcoin Core. Please do not edit this file while the node is running, as any changes might be ignored or overwritten."
```

Which seems a bit pointless/noisy, and could maybe even be confusing to users?

furszy commented:
> So now we will (always?) log this on startup:
> 
> ```shell
> 2024-01-23T22:48:12.205684Z [init] Setting file arg: _warning_ = "This file is automatically generated and updated by Bitcoin Core. Please do not edit this file while the node is running, as any changes might be ignored or overwritten."
> ```
> 
> Which seems a bit pointless/noisy, and could maybe even be confusing to users?

Agree. Good catch. Fixed in #29301.
